xtask/src/run.rs

```diff
@@ -83,40 +102,383 @@ pub fn build(opts: BuildOptions) -> Result<Vec<(String, PathBuf)>> {
+                    .with_context(|| format!("failed to wait for {gen_init_cpio:?}"))?;
+                let Output { status, .. } = &output;
+                if status.code() != Some(0) {
+                    bail!("{gen_init_cpio:?} failed: {output:?}")
+                }
+
+                copy(&initrd_image, "/tmp/initrd.img").context("copy failed")?;
+
+                let mut qemu = Command::new(format!("qemu-system-{guest_arch}"));
+                if let Some(machine) = machine {
+                    qemu.args(["-machine", machine]);
+                }
+                if guest_arch == ARCH {
+                    match OS {
+                        "linux" => match metadata("/dev/kvm") {
+                            Ok(metadata) => {
+                                use std::os::unix::fs::FileTypeExt as _;
+                                if metadata.file_type().is_char_device() {
+                                    qemu.args(["-accel", "kvm"]);
+                                }
+                            }
+                            Err(error) => {
+                                if error.kind() != ErrorKind::NotFound {
+                                    Err(error).context("failed to check existence of /dev/kvm")?;
+                                }
+                            }
+                        },
+                        "macos" => {
+                            qemu.args(["-accel", "hvf"]);
+                        }
+                        os => bail!("unsupported OS: {os}"),
+                    }
+                } else if let Some(cpu) = cpu {
+                    qemu.args(["-cpu", cpu]);
+                }
+                let console = OsString::from("ttyS0");
+                let kernel_args = std::iter::once(("console", &console))
+                    .chain(run_args.clone().map(|run_arg| ("init.arg", run_arg)))
+                    .enumerate()
+                    .fold(OsString::new(), |mut acc, (i, (k, v))| {
+                        if i != 0 {
+                            acc.push(" ");
+                        }
+                        acc.push(k);
+                        acc.push("=");
+                        acc.push(v);
+                        acc
+                    });
+                qemu.args(["-no-reboot", "-nographic", "-m", "512M", "-smp", "2"])
+                    .arg("-append")
+                    .arg(kernel_args)
+                    .arg("-kernel")
+                    .arg(&kernel_image)
+                    .arg("-initrd")
+                    .arg(&initrd_image);
+                if guest_arch == "aarch64" {
+                    match OS {
+                        "linux" => {
+                            let mut cmd = Command::new("locate");
+                            let output = cmd
+                                .arg("QEMU_EFI.fd")
+                                .output()
+                                .with_context(|| format!("failed to run {cmd:?}"))?;
```

Comment:
Will the kernel not boot without UEFI?
It would be a lot simpler in both cases (x86/aarch64) to live without it.

Reply:
Yes, it will not boot. Black screen and no progress.